### PR TITLE
add FSDP Shampoo optimizer with full-matrix AdaGrad computation

### PR DESCRIPTION
**Pull request by anana10c**
Create FSDP Shampoo optimizer based on DDP implementation, with the distributed buffer removed. All workers now compute gradient updates for all the parameters they store, without communicating with other workers during the optimizer step. As PyTorch FSDP flattens the parameters when sharded, this effectively computes blocked full-matrix AdaGrad when the sharding strategy is set to FULL_SHARD, and sequential Shampoo when it is set to NO_SHARD.

### CODE REVIEW

File: distributed_shampoo/fsdp_shampoo.py [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        iteration = self._iterate_step()
+        self._update_preconditioners()
+
+        # for debugging
+        current_rank = dist.get_rank()
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: Let's remember to remove these two lines of code once you're done debugging. :)

File: distributed_shampoo/fsdp_shampoo.py [hunk 2]
```diff
@@ -0,0 +1,849 @@
...
+
+            (
+                split_params,
+                split_preconditioned_grads,
+                split_momentum_directions,
+            ) = self._init_group(group)
+
+            for p in group[PARAMS]:
+                state = self.state[p]
+                if p.grad is None or not state[PRECONDITIONERS].on_source_rank:
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: We can remove the `not state[PRECONDITIONERS].on_source_rank` piece since we are no longer relying on any distributed computation. Do we also want to perform a consistent check to see if `p.numel() == 0` here?
```suggestion
                if p.numel() == 0 or p.grad is None:
```

File: distributed_shampoo/fsdp_shampoo.py [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+            else:
+                state[k] = v
+
+        # Update parameter groups, setting their 'params' value
+        def update_group(group, new_group):
+            new_group[PARAMS] = group[PARAMS]
+            return new_group
+
+        param_groups = [update_group(g, ng) for g, ng in zip(groups, saved_groups)]
+        self.__setstate__({"state": state, "param_groups": param_groups})
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: We will need to work with @awgu and Wanchao on how to support checkpointing in the future, both for open-source and internal workloads. In the meantime, let's just remove the `load_state_dict` function so there's less bloat. :)
```suggestion
```

File: distributed_shampoo/fsdp_shampoo.py [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+    def _compute_root_inverse(self):
+        """Root inverse computation across all preconditioners/parameters."""
+        for group in self.param_groups:
+            for p in group[PARAMS]:
+                # skip parameters not on worker
+                if p.numel() == 0:
+                    continue
+
+                if p.grad is None:
+                    continue
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: Let's combine these.
```suggestion
                # skip parameters not on worker
                if p.numel() == 0 or p.grad is None:
                    continue
```

File: distributed_shampoo/fsdp_shampoo.py [hunk 5]
```diff
@@ -0,0 +1,849 @@
...
+        self._world_size = dist.get_world_size() if self._use_distributed() else 1
+
+        # Initialize Shampoo preconditioners and distributed buffers.
+        self._dist_group = None
+        self._initialize_preconditioners_and_steps()
+
+    @torch.no_grad()
+    def _use_distributed(self) -> bool:
+        return self._num_trainers_per_group > 1
+
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: Let's remove this function here. :)
```suggestion
```

File: distributed_shampoo/fsdp_shampoo.py [hunk 6]
```diff
@@ -0,0 +1,849 @@
...
+            )
+        else:
+            self._start_preconditioning_step = start_preconditioning_step
+
+        # Initialize comms-related fields.
+        self._world_size = dist.get_world_size() if self._use_distributed() else 1
+
+        # Initialize Shampoo preconditioners and distributed buffers.
+        self._dist_group = None
+        self._initialize_preconditioners_and_steps()
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: Let's remove these variables here since they should not be used.
```suggestion
        self._initialize_preconditioners_and_steps()
```

File: distributed_shampoo/fsdp_shampoo.py [hunk 7]
```diff
@@ -0,0 +1,849 @@
...
+        use_decoupled_weight_decay: bool = True,
+        preconditioner_dtype: torch.dtype = torch.float,
+        large_dim_method: LargeDimMethod = LargeDimMethod.BLOCKING,
+        num_trainers_per_group: int = -1,
+        use_merge_dims: bool = True,
+        grafting_type: GraftingType = GraftingType.ADAGRAD,
+        grafting_epsilon: float = 1e-3,
+        grafting_beta2: float = 1.0,
+        use_protected_eigh: bool = True,
+        use_dtensor: bool = True,
```
Comment (hjmshi) [inline distributed_shampoo/fsdp_shampoo.py]: Let's ignore using `DTensor` for now and set this to `False`, unless you think we need it? :)

**Review by hjmshi — COMMENTED** [review]
Great work, @anana10c! Overall, very good. Just a few small nits. Let's finish up the initial enablement of something for Shampoo, and we can pivot to performing some real experiments!

File: distributed_shampoo/fsdp_shampoo.py (hunk 6 above)
Comment (anana10c) [reply distributed_shampoo/fsdp_shampoo.py]: I was thinking that we may need to use them in the future if we decide to experiment with communications, so I left them for now.

**Review by hjmshi — APPROVED** [review]
